test/src/_crcError_test.cpp

```diff
@@ -17,72 +17,175 @@
+                }
+            }
+
+            AND_GIVEN("The Notecard returns an error message") {
+                char json[] = "{\"err\":\"cannot interpret JSON: bool being placed into a non-bool field {io}\"}";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
+
+            AND_GIVEN("No CRC field")
+            {
+                char json[] = "{\"req\": \"hub.sync\"}";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
+
+            AND_GIVEN("The CRC field exists, but is not at the tail of the response")
+            {
+                char json[] = "{\"crc\":\"0009:10BAC79A\",\"req\": \"hub.sync\"}";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
         }
 
-        SECTION("Everything matches") {
-            char json[] = "{\"req\":\"hub.sync\"}";
-            char *jsonWithCrc = _crcAdd(json, seqNo);
-            REQUIRE(jsonWithCrc != NULL);
-
-            CHECK(!_crcError(jsonWithCrc, seqNo));
-
-            NoteFree(jsonWithCrc);
-        }
-
-        SECTION("Trailing CRLF") {
-            char json[] = "{\"req\":\"hub.sync\",\"crc\":\"0001:10BAC79A\"}\r\n";
-
-            // Trailing \r\n should be ignored.
-            CHECK(!_crcError(json, seqNo));
+        GIVEN("The Notecard firmware supports CRC") {
+            notecardFirmwareSupportsCrc = true;
+
+            NoteSetFnDefault(malloc, free, NULL, NULL);
+
+            uint16_t seqNo = 1;
+
+            AND_GIVEN("An empty string")
+            {
+                char json[] = "";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
+
+            AND_GIVEN("Invalid JSON")
+            {
+                char json[] = "{\"req\":";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
+
+            AND_GIVEN("The Notecard returns an error message") {
+                char json[] = "{\"err\":\"cannot interpret JSON: bool being placed into a non-bool field {io}\"}";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
+
+            AND_GIVEN("No CRC field")
+            {
+                char json[] = "{\"req\": \"hub.sync\"}";
+
+                THEN("A CRC error SHALL NOT be reported")
+                {
+                    CHECK(!_crcError(json, seqNo));
+                }
+            }
+
+            AND_GIVEN("The CRC field exists, but is not at the tail of the response")
```

Comment:
What is the difference between this test and line #73? Are they expected to report differently?

Reply:
The highest order branch (i.e. `GIVEN`), dictates whether `note-c` believes the Notecard firmware is capable of sending CRC's in the response.

- If the firmware does not support CRC, then a CRC is assumed to be of user origin and should not be tested.
- If the firmware does support CRC, but it's in the wrong position (then we assume it is of user origin and do not test it).